File: 每日一题/q2287. 重排字符形成目标字符串/2287.py
class Solution(object):
    def rearrangeCharacters(self, s, target):
        """
        :type s: str
        :type target: str
        :rtype: int
        """
        count = 0
        i = 0
        j = 0
        while i <= len(s)-len(target):
            if s[i] not in target:
                i += 1
            else:
                # print("The first item in ", i,"is ", s[i], "\n")
                Flag = self.check(i, s, target)
                if Flag:
                    i += len(target)
                    count += 1
                    # print("COUNT:",count,"\n")
                    # print("i is ", i,"\n")
                else:
                    i += 1
        return count

    def check(self, i, s, target):
        n_target = len(target)
        for j in range(n_target):
            if s[i+j] in target:
                target = target.replace(s[i+j], "")
                j += 1
            else:
                return False
        # print("The target from ", i, "is: ", s[i:i+n_target])
        return True


if __name__ == '__main__':
    sol1 = Solution()
    s1 = "ilovecodingonleetcode"
    target1 = "code"
    s2 = "abcba"
    target2 = "abc"
    s3 = "abbaccaddaeea"
    target3 = "aaaaa"

    print(sol1.rearrangeCharacters(s1, target1))

File: 每日一题/q2287. 重排字符形成目标字符串/2287.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
#include <climits>
using namespace std;

class Solution
{
public:
    int rearrangeCharacters_map(string s, string target)
    {
        unordered_map<char, int> sCounts, targetCounts;

        int n = s.size(), m = target.size();

        for (int i = 0; i < m; i++)
        {
            targetCounts[target[i]]++;
        }

        for (int i = 0; i < n; i++)
        {
            if (targetCounts.count(s[i]))
            {
                sCounts[s[i]]++;
            }
        }

        int ans = INT_MAX;

        for (auto &c : targetCounts)
        {
            // cout << "key:" << c.first << ", value:" << c.second << endl;
            int totalCount = sCounts[c.first];
            ans = min(ans, totalCount / c.second);
            if (ans == 0)
            {
                return 0;
            }
        }
        return ans;

        // for(auto &m : targetCounts){
        //     char key = *m.key();
        //     int value = *m.value();

        // }
    }

public:
    int rearrangeCharacters_vector(string s, string t)
    {
        vector<int> mp1(26, 0), mp2(26, 0);
        for (char &c : s)
            mp1[c - 'a']++;
        for (char &c : t)
            mp2[c - 'a']++;
        int ans = 105;
        for (char &c : t)
            ans = min(mp1[c - 'a'] / mp2[c - 'a'], ans);
        ;
        return ans;
    }
};

int main()
{
    Solution sol1;
    string s1 = string("ilovecodingonleetcode");
    string target1 = string("code");
    cout << sol1.rearrangeCharacters_map(s1, target1) << endl;
    return 0;
}